src/VisualStudio/Core/Def/Implementation/LanguageClient/AbstractInProcLanguageClient.cs

```diff
@@ -95,6 +99,31 @@ internal abstract class AbstractInProcLanguageClient : ILanguageClient
 
         public async Task<Connection?> ActivateAsync(CancellationToken cancellationToken)
         {
+            // HACK HACK HACK: prevent potential crashes/state corruption during load. Fixes
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1261421
+            //
+            // When we create an LSP server, we compute our server capabilities; this may depend on
+            // reading things like workspace options which will force us to initialize our option persisters.
+            // Unfortunately some of our option persisters currently assert they are first created on the UI
+            // thread. If the first time they're created is because of LSP initialization, we might end up loading
+            // them on a background thread which will throw exceptions and then prevent them from being created
+            // again later.
+            //
+            // The correct fix for this is to fix the threading violations in the option persister code;
+            // asserting a MEF component is constructed on the foreground thread is never allowed, but alas it's
+            // done there. Fixing that isn't difficult but comes with some risk I don't want to take for 16.9;
+            // instead we'll just compute our capabilities here on the UI thread to ensure everything is loaded.
+            // We _could_ consider doing a SwitchToMainThreadAsync in InProcLanguageServer.InitializeAsync
+            // (where the problematic call to GetCapabilites is), but that call is invoked across the StreamJsonRpc
+            // link where it's unclear if VS Threading rules apply. By doing this here, we are dong it in a
+            // VS API that is following VS Threading rules, and it also ensures that the preereqs are loaded
+            // prior to any RPC calls being made.
+            //
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/639090 will track removing this hack
+            // since that's the primary offending persister that needs to be addressed.
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+            _ = GetCapabilities();
+
```

Comment:
should we potentially switch back off the UI thread here, so that the creation of the connection, instantiating the LSP server can happen off of the uI thread?

Reply:
So it looks like this API was already being called on the UI thread; I agree conceptually that it's the right thing to do, but worried about added risk. Thoughts?

Reply:
Ah, if it's already on it then I'm OK with this - switching off would then be a new path (and potential for bugs).